src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,431 @@
+            }
+
+        } else if (hkdfType == HKDF_TYPE.EXPAND) {
+            // perform expand
+            try {
+                return Arrays.copyOf(hkdfExpand(this.pseudoRandomKey, this.info, this.length),
+                                     this.length);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF Expand could not be initialized with the given key material");
+            }
+
+        } else if (hkdfType == HKDF_TYPE.EXTRACTEXPAND) {
+            // perform extract and then expand
+            try {
+                byte[] extractResult = hkdfExtract(initialKeyMaterial,
+                                                   (salt == null) ? null : salt.getEncoded());
+                this.pseudoRandomKey = new SecretKeySpec(extractResult, "RAW");
+                return Arrays.copyOf(hkdfExpand(this.pseudoRandomKey, this.info, this.length),
+                                     this.length);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF ExtractExpand could not be initialized with the given key or salt "
+                    + "material");
+            }
+        }
+
+        throw new InvalidParameterSpecException(
+            "an HKDF could not be initialized with the given KDFParameterSpec");
+    }
+
+    protected int setupHMAC(String hmacAlgName) throws NoSuchAlgorithmException {
+        hmacObj = Mac.getInstance(hmacAlgName);
+        return hmacObj.getMacLength();
+    }
+
+    protected void inspectKDFParameterSpec(KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+        // A switch would be nicer, but we may need to backport this before JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable in instanceof either
+        if (kdfParameterSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract = (HKDFParameterSpec.Extract) kdfParameterSpec;
+            this.ikms = anExtract.ikms();
+            this.salts = anExtract.salts();
+            if (isNullOrEmpty(ikms) && isNullOrEmpty(salts)) {
+                throw new InvalidParameterSpecException(
+                    "IKM and salt cannot both be null or empty for HKDFParameterSpec.Extract");
+            }
+            // we should be able to combine these Lists of keys into single SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
+            try {
+                this.initialKeyMaterial = consolidateKeyMaterial(ikms);
+                this.salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "Issue encountered when combining ikm or salt values into single keys");
```

Comment:
add cause (ike) to exception.

Reply:
Unfortunately, in this case, you have to store the exception to a variable, call` setCause` on it, and then return it.